Improvements

webimpress opened the pull request:
- more generic tests in delegators to use all types of factories
- removed duplicated test as these are covered by more generic tests using all factory types
- refactor expected exception assertion
- added one more missing delegator test

webimpress commented:
With all these changes Aura.Di and Pimple integration PRs works:
- https://github.com/zendframework/zend-auradi-config/pull/10
- https://github.com/zendframework/zend-pimple-config/pull/10

weierophinney commented on src/AliasTestTrait.php:
```diff
@@ -13,7 +13,7 @@
 trait AliasTestTrait
 {
-    public function alias() : Generator
+    final public function alias() : Generator
```
Just noting for the record, in case anybody wonders about the `final` visibility: this is because we do not want implementors overriding these tests and testing different behavior.

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -63,31 +63,7 @@ public function testDelegatorsDoNotOperateOnServices() : void
         self::assertSame($myService, $instance);
     }
 
-    public function testDelegatorsOperateOnFactoryBackedServices() : void
```
Why is this test case removed?

webimpress replied:
Covered by `testDelegatorFactoriesTriggerForFactoryBackedServicesUsingAnyFactoryType` with data provider `factoriesForDelegators` (so more generic, for all types of factories).

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -143,67 +119,64 @@ public function testDelegatorsNamedForAliasDoNotApplyToInvokableServiceResolvedV
     }
 
-    public function testDelegatorsDoNotApplyToAliasResolvingToServiceEntry() : void
+    final public function testDelegatorsNamedForAliasDoNotApplyToInvokableServiceWithAlias() : void
```
Why are these two tests swapped? It makes understanding the diff far harder (I had to look back and forth around a half-dozen times to realize that was what happened...)

webimpress replied:
Diff is messed up. Test `testDelegatorsNamedForAliasDoNotApplyToInvokableServiceWithAlias` was added here, as it is similar to the previous so it was logic place to add it. 

Test `testDelegatorsDoNotApplyToAliasResolvingToServiceEntry` has not been modified at all, just added `final`.

weierophinney replied:
`testDelegatorsDoNotApplyToAliasResolvingToServiceEntry` was moved; it now appears after the new test method.

webimpress replied:
It was not moved, another one was added before it, if you check "split view" maybe it will be clearer.

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -231,39 +204,7 @@ public function testDelegatorsDoNotTriggerForAliasTargetingInvokableService() :
         self::assertSame($instance, $container->get(TestAsset\Service::class));
     }
 
-    public function testDelegatorsTriggerForFactoryServiceResolvedByAlias() : void
```
I found where this was moved, and discovered you added a data provider, which is excellent... but why was it moved? (Moving things around makes code review sooooo hard!)

webimpress replied:
It was moved to keep logical order, I've used dataProvders to provide more generic tests for all factory types.

I agree that moving things around makes code review harder. I think it was needed here to keep it clear in code later on.

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -342,7 +280,32 @@ public function testDelegatorsReceiveCallbackResolvingToReturnValueOfPrevious(
         self::assertSame($instance, $container->get($serviceNameToTest));
     }
 
-    public function testMultipleAliasesForADelegatedInvokableServiceReceiveSameInstance()
```
Again: why was this test moved?

webimpress replied:
It was not moved, just diff is show it wrong, one test was added before that one, and here only `final` keyword was added.

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -480,205 +448,184 @@ public function testDelegatorFactoriesTriggerForFactoryBackedServicesUsingAnyFac
     }
-    
-    // @codingStandardsIgnoreStart
-    public function testWhenInvokableWithDelegatorsResolvesToInvalidClassAnExceptionIsRaisedIfCallbackNeverInvoked()
```
Why was this test removed?

webimpress replied:
Covered by more generic test `testWithDelegatorsResolvesToInvalidClassNoExceptionIsRaisedIfCallbackNeverInvoked` with data provider `invalidService`.

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -480,205 +448,184 @@ public function testDelegatorFactoriesTriggerForFactoryBackedServicesUsingAnyFac
-        self::assertInstanceOf(TestAsset\Delegator::class, $instance);
-    }
-    
-    public function testWhenInvokableWithDelegatorsResolvesToNonExistentClassAnExceptionIsRaisedWhenCallbackIsInvoked()
```
Why was this test removed?

webimpress replied:
Covered by more generic test `testWithDelegatorsResolvesToInvalidClassAnExceptionIsRaisedWhenCallbackIsInvoked` with data provider `invalidService`.

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -480,205 +448,184 @@ public function testDelegatorFactoriesTriggerForFactoryBackedServicesUsingAnyFac
     }
-    
-    // @codingStandardsIgnoreStart
-    public function testWhenInvokableWithDelegatorsResolvesToInvalidFactoryClassAnExceptionIsRaisedWhenCallbackIsInvoked()
```
Why was this test removed?

webimpress replied:
Covered by more generic test `testWithDelegatorsResolvesToInvalidClassAnExceptionIsRaisedWhenCallbackIsInvoked` with data provider `invalidService`.

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -480,205 +448,184 @@ public function testDelegatorFactoriesTriggerForFactoryBackedServicesUsingAnyFac
-    }
-    
-    // @codingStandardsIgnoreStart
-    public function testWhenServiceWithDelegatorsResolvesToNonExistentFactoryClassNoExceptionIsRaisedIfCallbackNeverInvoked()
```
Why was this test removed?

webimpress replied:
Covered by `testWithDelegatorsResolvesToInvalidClassNoExceptionIsRaisedIfCallbackNeverInvoked` with data provider `invalidService`.

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -480,205 +448,184 @@ public function testDelegatorFactoriesTriggerForFactoryBackedServicesUsingAnyFac
-    }
-    
-    // @codingStandardsIgnoreStart
-    public function testWhenServiceWithDelegatorsResolvesToInvalidFactoryClassAnExceptionIsRaisedIfCallbackNeverInvoked()
```
Why was this test removed?

webimpress replied:
The test name was wrong, it was testing that "__no__ exception is raised ...".

Covered by `testWithDelegatorsResolvesToInvalidClassNoExceptionIsRaisedIfCallbackNeverInvoked` with data provider `invalidService`

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -480,205 +448,184 @@ public function testDelegatorFactoriesTriggerForFactoryBackedServicesUsingAnyFac
     }
-    
-    // @codingStandardsIgnoreStart
-    public function testWhenServiceWithDelegatorsResolvesToNonExistentFactoryClassAnExceptionIsRaisedWhenCallbackIsInvoked()
```
Why was this test removed?

webimpress replied:
Covered by `testWithDelegatorsResolvesToInvalidClassAnExceptionIsRaisedWhenCallbackIsInvoked` with data provider `invalidService`.

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -480,205 +448,184 @@ public function testDelegatorFactoriesTriggerForFactoryBackedServicesUsingAnyFac
-        $container->get('service');
-    }
-    
-    public function testWhenServiceWithDelegatorsResolvesToInvalidFactoryClassAnExceptionIsRaisedWhenCallbackIsInvoked()
```
Why was this test removed?

webimpress replied:
Covered by `testWithDelegatorsResolvesToInvalidClassAnExceptionIsRaisedWhenCallbackIsInvoked` with data provider `invalidService`

webimpress commented:
@weierophinney I think the diff is messed up somehow, it's better to have a look on separate commits. Basically I've added data providers to have more cases and removed duplicated tests (as data providers covers all cases). I can write more explanations tomorrow, but I think when you have a look commit by commit it should be clearer.

webimpress commented on src/DelegatorTestTrait.php:
```diff
@@ -143,67 +119,64 @@ public function testDelegatorsNamedForAliasDoNotApplyToInvokableServiceResolvedV
+        self::assertSame($instance, $container->get(TestAsset\Service::class));
     }
 
-    public function testDelegatorsDoNotTriggerForAliasTargetingFactoryBasedService() : void
```
This one is removed, because it's covered by `testDelegatorsDoNotTriggerForAliasTargetingFactoryBasedServiceUsingAnyFactoryType`.

webimpress commented on src/DelegatorTestTrait.php:
```diff
@@ -281,14 +222,6 @@ public function delegatorService() : Generator
             TestAsset\Service::class,
         ];
 
-        yield 'factory-service' => [
```
Check `foreach` loop below - we are generating there entities for all factory types, the same about `alias-of-factory-service`.
